metaflow/plugins/kfp/kfp.py

```diff
@@ -1,316 +1,419 @@
-                                                                     run_name=run_name,
-                                                                     namespace=namespace)
-    return run_pipeline_result
+        cmds = []
 
+        script_name = os.path.basename(sys.argv[0])
+        executable = self.environment.executable(node.name)
 
-def create_kfp_pipeline_yaml(flow_graph, code_url, pipeline_file_path=DEFAULT_KFP_YAML_OUTPUT_PATH):
+        if R.use_r():
+            entrypoint = [R.entrypoint()]
+        else:
+            entrypoint = [executable, script_name]
+
+        paths = input_paths
+        kfp_run_id = "kfp-" + dsl.RUN_ID_PLACEHOLDER
+
+        # TODO: use dynamic task_id
+        #   But then we must save it in Dynamo (a kv store) for joins
+        #   task_id = "{{pod.name}}"  # ARGO pod name
+        if node.name == "start":
+            # We need a separate unique ID for the special _parameters task
+            task_id_params = "%s-params" % task_id
+            # Export user-defined parameters into runtime environment
+            param_file = "".join(
+                random.choice(string.ascii_lowercase) for _ in range(10)
+            )
+
+            # TODO: move to KFP plugin
+            export_params = (
+                "python -m "
+                "metaflow.plugins.aws.step_functions.set_batch_environment "
+                "parameters %s && . `pwd`/%s" % (param_file, param_file)
+            )
+            params = entrypoint + [
+                "--quiet",
+                "--metadata=%s" % self.metadata.TYPE,
+                "--environment=%s" % self.environment.TYPE,
+                "--datastore=s3",
+                "--datastore-root=%s" % self.datastore.datastore_root,
+                "--event-logger=%s" % self.event_logger.logger_type,
+                "--monitor=%s" % self.monitor.monitor_type,
+                "--no-pylint",
+                "init",
+                "--run-id %s" % kfp_run_id,
+                "--task-id %s" % task_id_params,
+            ]
+
+            # If the start step gets retried, we must be careful not to
+            # regenerate multiple parameters tasks. Hence we check first if
+            # _parameters exists already.
+            task_id_params_path = "{kfp_run_id}/_parameters/{task_id_params}".format(
+                kfp_run_id=kfp_run_id, task_id_params=task_id_params
+            )
+            exists = entrypoint + ["dump", "--max-value-size=0", task_id_params_path]
+            cmd = "if ! %s >/dev/null 2>/dev/null; then %s && %s; fi" % (
+                " ".join(exists),
+                export_params,
+                " ".join(params),
+            )
+            cmds.append(cmd)
+            paths = task_id_params_path
+
+        if node.type == "join" and self.graph[node.split_parents[-1]].type == "foreach":
+            # TODO: get from Dynamo or a kv store
+            parent_tasks_file = "".join(
```

Comment:
What is `parent_tasks_file`? I'd add an inline comment.

Reply:
removed